generics/fanout.go

```diff
@@ -0,0 +1,163 @@
+package generics
+
+import "sync"
+
+// Fanout takes a slice of input, a parallelism factor, and a worker factory. It
+// calls the generated worker on every element of the input, and returns a
+// (possibly filtered) slice of the outputs in no particular order. Only the
+// outputs that pass the predicate (if it is not nil) will be added to the
+// slice.
+//
+// The factory takes an integer (the worker number) and constructs a function of
+// type func(T) U that processes a single input and produces a single output. It
+// also constructs a cleanup function, which may be nil. The cleanup function is
+// called once for each worker, after the worker has completed processing all of
+// its inputs. It is given the same index as the corresponding worker factory.
+//
+// If predicate is not nil, it will only add the output to the result slice if
+// the predicate returns true. It will fan out the input to the worker function
+// in parallel, and fan in the results to the output slice.
+func Fanout[T, U any](input []T, parallelism int, workerFactory func(int) (worker func(T) U, cleanup func(int)), predicate func(U) bool) []U {
```

Comment:
I don't see this function used anywhere, should we hold off adding it until we have a use case?

Reply:
It's not yet, but this is in a generics folder intended for broader use, AND I expect to find other uses for it within Refinery.